include/zephyr/net/coap_client.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_NET_COAP_CLIENT_H_
+#define ZEPHYR_INCLUDE_NET_COAP_CLIENT_H_
+
+#include <zephyr/net/coap.h>
+
+
+#define MAX_COAP_MSG_LEN CONFIG_COAP_ASYNC_CLIENT_BUFFER_SIZE
+
+/**
+ * @typedef coap_client_response_cb_t
+ * @brief Callback for CoAP request.
+ *
+ * This callback is called for responses to CoAP client requests.
+ * It is used to indicate errors, response codes from server or to deliver payload.
+ * Blockwise transfers cause this callback to be called sequentially with increasing payload offset
+ * and only partial content in buffer pointed by payload parameter.
+ *
+ * @param result_code Result code of the response. Negative if there was a failure in send.
+ *                    @ref coap_response_code for positive.
+ * @param offset Payload offset from the beginning of a blockwise transfer.
+ * @param payload Buffer containing the payload from the response. NULL for empty payload.
+ * @param len Size of the payload.
+ * @param last_block Indicates the last block of the response.
+ * @param user_data User provided context.
+ */
+typedef void (*coap_client_response_cb_t)(uint8_t result_code,
+					  size_t offset, const uint8_t *payload, size_t len,
+					  bool last_block, void *user_data);
+
+/**
+ * @brief Representation of a CoAP client request.
+ */
+struct coap_client_request {
+	enum coap_method method;      /**< Method of the request */
+	bool confirmable;	      /**< CoAP Confirmable/Non-confirmable message */
+	const char *path;	      /**< Path of the requested resource */
+	enum coap_content_format fmt; /**< Content format to be used */
```

Comment:
This enum does not have a field for "ignore" or "unknown" so having it here forces Accept option to be always in the query.
In my mind, this feels a bit overkill. I have not yet seen a CoAP service that uses Accept option to change the content format. Most have hard coded.

Is this really necessary?

Reply:
It may be overkill in some sense, but the standard does allow for it.  I know there is no point in implementing every possible option though.

That's a good point that there is no ignore or unknown value.  It would be nice to be able to omit either Accept or Content-Type.  Perhaps add such a value to the enum?  Or add another value to control which options to include, for example, a bitmask?

Using Accept on responses is currently necessary for the server we are developing, but I will discuss offline and respond later.

Reply:
Added 2 new fields. A list of extra options and a number of options accompanying it. This should now satisfy also the possible future requirements for extra options. Also added a KConfig for defining header size. If adding lots of extra headers, users should increase the header size as well to make sure the header and payload will fit to the buffer.